### PR TITLE
Use shard and replica name from `Replicated` database arguments

### LINKED ISSUE #31471: Pass shard and replica from DatabaseReplicated to create table
**Describe the issue**
We have to set macro or to pass explicitly shard and replica name to create table in DatabaseReplicated

### PR DESCRIPTION
**Pull request by tavplubix**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Use shard and replica name from `Replicated` database arguments when expanding macros in `ReplicatedMergeTree` arguments if these macros are not defined in config. Closes #31471.

### CODE REVIEW

File: tests/integration/test_replicated_database/test.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 dummy_node = cluster.add_instance('dummy_node', main_configs=['configs/config.xml'], user_configs=['configs/settings.xml'], with_zookeeper=True, stay_alive=True, macros={"shard": 1, "replica": 2})
 competing_node = cluster.add_instance('competing_node', main_configs=['configs/config.xml'], user_configs=['configs/settings.xml'], with_zookeeper=True, macros={"shard": 1, "replica": 3})
 snapshotting_node = cluster.add_instance('snapshotting_node', main_configs=['configs/config.xml'], user_configs=['configs/settings.xml'], with_zookeeper=True, macros={"shard": 2, "replica": 1})
-snapshot_recovering_node = cluster.add_instance('snapshot_recovering_node', main_configs=['configs/config.xml'], user_configs=['configs/settings.xml'], with_zookeeper=True, macros={"shard": 2, "replica": 2})
+snapshot_recovering_node = cluster.add_instance('snapshot_recovering_node', main_configs=['configs/config.xml'], user_configs=['configs/settings.xml'], with_zookeeper=True)
```
Comment (vdimir) [inline tests/integration/test_replicated_database/test.py:19]: I didn't get if this test actually test changes. 
Only `snapshot_recovering_node` do not define macros, but I don't see queries that will check that `snapshot_recovering_node` have correct shard and replica name.
Could you clarify it?

**Review by vdimir — APPROVED** [review]

File: tests/integration/test_replicated_database/test.py (hunk 1 above)
Comment (tavplubix) [reply tests/integration/test_replicated_database/test.py:19]: Replicated table is created on `main_node` [here](https://github.com/ClickHouse/ClickHouse/blob/2f8b8596a243fc99777113884d20a5ac9c789e7f/tests/integration/test_replicated_database/test.py#L269-L271). Then we create new database replica on `snapshot_recovering_node` and check that it successfully created the table [here](https://github.com/ClickHouse/ClickHouse/blob/2f8b8596a243fc99777113884d20a5ac9c789e7f/tests/integration/test_replicated_database/test.py#L284-L292).

**Comment by tavplubix** [comment]
ClickHouse build check - something strange:
```
2021-11-18 10:14:17 [4122/5811] Building CXX object src/CMakeFiles/dbms.dir/AggregateFunctions/AggregateFunctionFactory.cpp.o
2021-11-18 10:14:17 In file included from /usr/include/c++/11/string:40,
2021-11-18 10:14:17                  from /usr/include/c++/11/stdexcept:39,
2021-11-18 10:14:17                  from /usr/include/c++/11/system_error:41,
2021-11-18 10:14:17                  from /usr/include/c++/11/bits/std_mutex.h:39,
2021-11-18 10:14:17                  from /usr/include/c++/11/bits/atomic_wait.h:49,
2021-11-18 10:14:17                  from /usr/include/c++/11/bits/atomic_base.h:41,
2021-11-18 10:14:17                  from /usr/include/c++/11/atomic:41,
2021-11-18 10:14:17                  from /usr/include/boost/smart_ptr/detail/atomic_count_std_atomic.hpp:16,
2021-11-18 10:14:17                  from /usr/include/boost/smart_ptr/detail/atomic_count.hpp:77,
2021-11-18 10:14:17                  from /usr/include/boost/smart_ptr/intrusive_ref_counter.hpp:19,
2021-11-18 10:14:17                  from ../src/Common/COW.h:4,
2021-11-18 10:14:17                  from ../src/Columns/IColumn.h:3,
2021-11-18 10:14:17                  from ../src/Columns/ColumnTuple.h:3,
2021-11-18 10:14:17                  from ../src/AggregateFunctions/IAggregateFunction.h:3,
2021-11-18 10:14:17                  from ../src/AggregateFunctions/AggregateFunctionFactory.h:3,
2021-11-18 10:14:17                  from ../src/AggregateFunctions/AggregateFunctionFactory.cpp:1:
2021-11-18 10:14:17 In static member function 'static constexpr int std::char_traits<char>::compare(const char_type*, const char_type*, std::size_t)',
2021-11-18 10:14:17     inlined from 'constexpr int std::basic_string_view<_CharT, _Traits>::compare(std::basic_string_view<_CharT, _Traits>) const [with _CharT = char; _Traits = std::char_traits<char>]' at /usr/include/c++/11/string_view:315:34,
2021-11-18 10:14:17     inlined from 'constexpr int std::basic_string_view<_CharT, _Traits>::compare(std::basic_string_view<_CharT, _Traits>::size_type, std::basic_string_view<_CharT, _Traits>::size_type, std::basic_string_view<_CharT, _Traits>) const [with _CharT = char; _Traits = std::char_traits<char>]' at /usr/include/c++/11/string_view:323:50,
2021-11-18 10:14:17     inlined from 'constexpr bool std::basic_string_view<_CharT, _Traits>::ends_with(std::basic_string_view<_CharT, _Traits>) const [with _CharT = char; _Traits = std::char_traits<char>]' at /usr/include/c++/11/string_view:366:22,
2021-11-18 10:14:17     inlined from 'bool std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::ends_with(std::basic_string_view<_CharT, _Traits>) const [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]' at /usr/include/c++/11/bits/basic_string.h:3065:63,
2021-11-18 10:14:17     inlined from 'DB::AggregateFunctionPtr DB::AggregateFunctionFactory::getImpl(const String&, const DataTypes&, const DB::Array&, DB::AggregateFunctionProperties&, bool) const' at ../src/AggregateFunctions/AggregateFunctionFactory.cpp:181:68:
2021-11-18 10:14:17 /usr/include/c++/11/bits/char_traits.h:361:32: warning: 'int __builtin_memcmp_eq(const void*, const void*, long unsigned int)' specified bound 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Wstringop-overread]
2021-11-18 10:14:17   361 |         return __builtin_memcmp(__s1, __s2, __n);
2021-11-18 10:14:17       |                ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
2021-11-18 10:14:17 In file included from /usr/include/c++/11/string:55,
2021-11-18 10:14:17                  from /usr/include/c++/11/stdexcept:39,
2021-11-18 10:14:17                  from /usr/include/c++/11/system_error:41,
2021-11-18 10:14:17                  from /usr/include/c++/11/bits/std_mutex.h:39,
2021-11-18 10:14:17                  from /usr/include/c++/11/bits/atomic_wait.h:49,
2021-11-18 10:14:17                  from /usr/include/c++/11/bits/atomic_base.h:41,
2021-11-18 10:14:17                  from /usr/include/c++/11/atomic:41,
2021-11-18 10:14:17                  from /usr/include/boost/smart_ptr/detail/atomic_count_std_atomic.hpp:16,
2021-11-18 10:14:17                  from /usr/include/boost/smart_ptr/detail/atomic_count.hpp:77,
2021-11-18 10:14:17                  from /usr/include/boost/smart_ptr/intrusive_ref_counter.hpp:19,
2021-11-18 10:14:17                  from ../src/Common/COW.h:4,
2021-11-18 10:14:17                  from ../src/Columns/IColumn.h:3,
2021-11-18 10:14:17                  from ../src/Columns/ColumnTuple.h:3,
2021-11-18 10:14:17                  from ../src/AggregateFunctions/IAggregateFunction.h:3,
2021-11-18 10:14:17                  from ../src/AggregateFunctions/AggregateFunctionFactory.h:3,
2021-11-18 10:14:17                  from ../src/AggregateFunctions/AggregateFunctionFactory.cpp:1:
2021-11-18 10:14:17 /usr/include/c++/11/bits/basic_string.h: In member function 'DB::AggregateFunctionPtr DB::AggregateFunctionFactory::getImpl(const String&, const DataTypes&, const DB::Array&, DB::AggregateFunctionProperties&, bool) const':
2021-11-18 10:14:17 /usr/include/c++/11/bits/basic_string.h:187:28: note: source object allocated here
2021-11-18 10:14:17   187 |       { return _M_dataplus._M_p; }

```

**Comment by tavplubix** [comment]
Anyway, let's restart
@Mergifyio update

**Comment by tavplubix** [comment]
Integration tests (release) - `test_replicated_merge_tree_hdfs_zero_copy/test.py::test_hdfs_zero_copy_with_ttl_delete`, cc: @yuzhichang, @kssenii

**Pull request merged**